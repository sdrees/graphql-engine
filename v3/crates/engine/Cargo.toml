[package]
name = "engine"
version.workspace = true
edition.workspace = true
license.workspace = true

build = "build.rs"

[lib]
bench = false

[[bin]]
name = "engine"
path = "bin/engine/main.rs"
bench = false

[[bench]]
name = "execute"
harness = false

[dependencies]
execute = { path = "../execute" }
hasura-authn-core = { path = "../auth/hasura-authn-core" }
hasura-authn-jwt = { path = "../auth/hasura-authn-jwt" }
hasura-authn-webhook = { path = "../auth/hasura-authn-webhook" }
lang-graphql = { path = "../lang-graphql" }
open-dds = { path = "../open-dds" }
opendds-derive = { path = "../utils/opendds-derive" }
schema = { path = "../schema" }
tracing-util = { path = "../utils/tracing-util" }
metadata-resolve = {path = "../metadata-resolve" }

anyhow = "1"
axum = { version = "0.6.20" }
base64 = "0.22.1"
bincode = "1.3.3"
clap = { version = "4", features = ["derive", "env"] }
json_value_merge = "2.0"
reqwest = { version = "0.11", features = ["json", "multipart"] }
schemars = { version = "0.8.20", features = ["smol_str"] }
serde = "1.0.202"
serde_json = "1.0.116"
serde_path_to_error = "0.1.14"
thiserror = "1.0"
tokio = { version = "1.26.0", features = [
  "macros",
  "parking_lot",
  "rt-multi-thread",
  "signal",
] }
tower = "0.4"
tower-http = { version = "0.4", features = ["trace", "cors", "fs"] }

# Dependencies to build and run build.rs file
[build-dependencies]
build-data = "0.1.5" # To set short commit-sha at build time

[dev-dependencies]
criterion = { version = "0.5", features = ["html_reports", "async_tokio"] }
goldenfile = "1.7.1"
pretty_assertions = "1.3.0"
tokio-test = "0.4.2"

[package.metadata.cargo-machete]
ignored = [
  "build_data",          # cargo-machete ignores build dependencies
  "serde_path_to_error", # used by opendds-derive
]

[lints]
workspace = true
