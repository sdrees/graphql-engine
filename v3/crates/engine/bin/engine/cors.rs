use axum::Router;
use reqwest::Method;
use std::time::Duration;
use tower_http::cors;

/// Add CORS layer to the app.
pub fn add_cors_layer(app: Router, cors_allow_origin: &[String]) -> Router {
    let cors_allow_origin = if cors_allow_origin.is_empty() {
        // Allow all origins and mirror the request origin in 'Access-Control-Allow-Origin'
        cors::AllowOrigin::mirror_request()
    } else {
        let allowed_origins = cors_allow_origin.to_owned();
        cors::AllowOrigin::predicate(move |origin_header_value, _req| {
            allowed_origins.iter().any(|allowed_origin| {
                // The allowed origins can include leading whitespace characters when
                // provided as a comma-space-separated list.
                // Example: --cors-allow-origin = 'val1, val2, val3'
                origin_header_value == allowed_origin.trim()
            })
        })
    };

    let cors = cors::CorsLayer::new()
        .max_age(Duration::from_secs(24 * 60 * 60)) // 24 hours
        .allow_headers(cors::AllowHeaders::mirror_request())
        .allow_origin(cors_allow_origin)
        .allow_credentials(true)
        .allow_methods(vec![Method::GET, Method::POST, Method::OPTIONS]);

    app.layer(cors)
}

#[cfg(test)]
mod test {
    use axum::{
        body::Body,
        http::{HeaderValue, Request},
        Router,
    };
    use pretty_assertions::assert_eq;
    use reqwest::StatusCode;
    use tower::ServiceExt;

    #[tokio::test]
    async fn test_no_cors() {
        let app = Router::new();
        // Preflight CORS request
        let response = app
            .oneshot(
                Request::builder()
                    .uri("/")
                    .method("OPTIONS")
                    .header("Origin", "http://example.com")
                    .body(Body::empty())
                    .unwrap(),
            )
            .await
            .unwrap();

        assert_eq!(response.status(), StatusCode::NOT_FOUND);
    }

    #[tokio::test]
    async fn test_cors_allow_all_origins() {
        let app = Router::new();
        let app = super::add_cors_layer(app, &[]);
        // Preflight CORS request
        let response = app
            .oneshot(
                Request::builder()
                    .uri("/")
                    .method("OPTIONS")
                    .header("Origin", "http://example.com")
                    .body(Body::empty())
                    .unwrap(),
            )
            .await
            .unwrap();

        assert_eq!(response.status(), StatusCode::OK);

        assert_eq!(
            response.headers().get("access-control-allow-origin"),
            Some(&HeaderValue::from_static("http://example.com"))
        );
    }

    #[tokio::test]
    async fn test_cors_restrict_origin() {
        let app = Router::new();
        let app = super::add_cors_layer(app, &["http://example.com".to_string()]);
        // Preflight CORS request
        let response = app
            .oneshot(
                Request::builder()
                    .uri("/")
                    .method("OPTIONS")
                    .header("Origin", "http://localhost:8080") // Only http://example.com is allowed origin
                    .body(Body::empty())
                    .unwrap(),
            )
            .await
            .unwrap();

        assert_eq!(response.status(), StatusCode::OK);

        assert_eq!(
            response.headers().get("access-control-allow-origin"), // Response shouldn't contain the allow origin header
            None
        );
    }

    #[tokio::test]
    async fn test_cors_allow_origin() {
        let app = Router::new();
        let app = super::add_cors_layer(
            app,
            &[
                "http://localhost:8080".to_string(),
                "http://example.com".to_string(),
            ],
        );
        // Preflight CORS request
        let response = app
            .oneshot(
                Request::builder()
                    .uri("/")
                    .method("OPTIONS")
                    .header("Origin", "http://example.com")
                    .body(Body::empty())
                    .unwrap(),
            )
            .await
            .unwrap();

        assert_eq!(response.status(), StatusCode::OK);

        assert_eq!(
            response.headers().get("access-control-allow-origin"),
            Some(&HeaderValue::from_static("http://example.com"))
        );
    }
}
