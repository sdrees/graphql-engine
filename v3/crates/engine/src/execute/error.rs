use gql::{ast::common as ast, http::GraphQLError};
use lang_graphql as gql;
use reqwest::StatusCode;
use serde_json as json;
use thiserror::Error;
use tracing_util::{ErrorVisibility, TraceableError};
use transitive::Transitive;

use crate::execute::ndc::client as ndc_client;

use super::ir;
use super::plan;
use super::types::Annotation;

/// Request errors are raised before execution of root fields begins.
/// Ref: https://spec.graphql.org/October2021/#sec-Errors.Request-errors
#[derive(Error, Debug)]
pub enum RequestError {
    #[error("parsing failed: {0}")]
    ParseFailure(#[from] gql::ast::spanning::Positioned<gql::parser::Error>),

    #[error("validation failed: {0}")]
    ValidationFailed(#[from] gql::validation::Error),

    #[error("{0}")]
    IRConversionError(#[from] ir::error::Error),

    #[error("error while generating plan: {0}")]
    PlanError(#[from] plan::error::Error),

    #[error("explain error: {0}")]
    ExplainError(String),
}

impl RequestError {
    pub fn to_graphql_error(self) -> GraphQLError {
        let message = match self {
            // Error messages for internal errors from IR conversion and Plan generations are masked.
            Self::IRConversionError(ir::error::Error::InternalError(_))
            | Self::PlanError(plan::error::Error::Internal(_)) => "internal error".into(),
            e => e.to_string(),
        };
        GraphQLError {
            message,
            path: None,
            extensions: None,
        }
    }
}

impl TraceableError for RequestError {
    fn visibility(&self) -> ErrorVisibility {
        match self {
            Self::IRConversionError(ir_error) => ir_error.visibility(),
            Self::PlanError(plan_error) => plan_error.visibility(),
            // Rest all errors are visible to users via traces
            Self::ParseFailure(_) | Self::ValidationFailed(_) | Self::ExplainError(_) => {
                ErrorVisibility::User
            }
        }
    }
}

/// Field errors are raised during execution from a root field
/// Ref: https://spec.graphql.org/October2021/#sec-Errors.Field-errors
#[derive(Error, Debug, Transitive)]
#[transitive(from(json::Error, FieldInternalError))]
#[transitive(from(NDCUnexpectedError, FieldInternalError))]
#[transitive(from(gql::normalized_ast::Error, FieldInternalError))]
#[transitive(from(gql::introspection::Error, FieldInternalError))]
pub enum FieldError {
    #[error("error from data source: {}", connector_error.error_response.message)]
    NDCExpected {
        connector_error: ndc_client::ConnectorError,
    },

    #[error("field '{field_name:} not found in _Service")]
    FieldNotFoundInService { field_name: String },

    #[error("internal error: {0}")]
    InternalError(#[from] FieldInternalError),
}

impl FieldError {
    fn get_details(&self) -> Option<serde_json::Value> {
        match self {
            Self::NDCExpected { connector_error } => {
                Some(connector_error.error_response.details.clone())
            }
            Self::InternalError(internal) => internal.get_details(),
            _ => None,
        }
    }

    pub fn to_graphql_error(self, path: Option<Vec<gql::http::PathSegment>>) -> GraphQLError {
        let details = self.get_details();
        match self {
            Self::InternalError(_internal) => GraphQLError {
                message: "internal error".into(),
                path,
                // Internal errors showing up in the API response is not desirable.
                // Hence, extensions are masked for internal errors.
                extensions: None,
            },
            e => GraphQLError {
                message: e.to_string(),
                path,
                extensions: details.map(|details| gql::http::Extensions { details }),
            },
        }
    }
}

impl TraceableError for FieldError {
    fn visibility(&self) -> ErrorVisibility {
        match self {
            Self::NDCExpected { .. } | Self::FieldNotFoundInService { .. } => ErrorVisibility::User,
            Self::InternalError(internal_error) => internal_error.visibility(),
        }
    }
}

#[derive(Error, Debug)]
pub enum FieldInternalError {
    #[error("ndc_unexpected: {0}")]
    NDCUnexpected(#[from] NDCUnexpectedError),

    #[error("Mapping for the Global ID typename {type_name:} not found")]
    GlobalIdTypenameMappingNotFound { type_name: ast::TypeName },

    #[error("introspection error: {0}")]
    IntrospectionError(#[from] gql::introspection::Error),

    #[error("error from normalized AST: {0}")]
    NormalizedAstError(#[from] gql::normalized_ast::Error),

    #[error("JSON serialization error: {0}")]
    JsonSerialization(#[from] json::Error),

    #[error("unexpected annotation: {annotation}")]
    UnexpectedAnnotation { annotation: Annotation },

    #[error("internal error: {description}")]
    InternalGeneric { description: String },
}

impl FieldInternalError {
    pub fn get_details(&self) -> Option<serde_json::Value> {
        match self {
            Self::NDCUnexpected(ndc_unexpected) => ndc_unexpected.get_details(),
            _ => None,
        }
    }
}

impl TraceableError for FieldInternalError {
    fn visibility(&self) -> ErrorVisibility {
        match self {
            Self::NDCUnexpected(_) | Self::GlobalIdTypenameMappingNotFound { .. } => {
                ErrorVisibility::User
            }
            Self::UnexpectedAnnotation { .. }
            | Self::JsonSerialization(_)
            | Self::InternalGeneric { .. }
            | Self::NormalizedAstError(_)
            | Self::IntrospectionError(_) => ErrorVisibility::Internal,
        }
    }
}

#[derive(Error, Debug)]
pub enum NDCUnexpectedError {
    #[error("ndc_client error: {0}")]
    NDCClientError(ndc_client::Error),

    #[error("unexpected response from data connector: {summary}")]
    BadNDCResponse { summary: String },
}

impl NDCUnexpectedError {
    pub fn get_details(&self) -> Option<serde_json::Value> {
        match self {
            Self::NDCClientError(ndc_client::Error::ConnectorError(ce)) => {
                Some(ce.error_response.details.clone())
            }
            _ => None,
        }
    }
}

// Convert NDC errors
impl From<ndc_client::Error> for FieldError {
    fn from(ndc_client_error: ndc_client::Error) -> FieldError {
        if let ndc_client::Error::ConnectorError(err) = &ndc_client_error {
            if matches!(
                err.status,
                // We forward the errors with status code 200 (OK), 403(FORBIDDEN), 409(CONFLICT) and 422(UNPROCESSABLE_ENTITY)
                StatusCode::OK
                    | StatusCode::FORBIDDEN
                    | StatusCode::CONFLICT
                    | StatusCode::UNPROCESSABLE_ENTITY
            ) {
                return FieldError::NDCExpected {
                    connector_error: err.clone(),
                };
            }
        }
        FieldError::InternalError(FieldInternalError::NDCUnexpected(
            NDCUnexpectedError::NDCClientError(ndc_client_error),
        ))
    }
}
