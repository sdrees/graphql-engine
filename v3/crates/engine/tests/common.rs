use goldenfile::{differs::text_diff, Mint};
use hasura_authn_core::{Identity, Role, Session, SessionError, SessionVariableValue};
use lang_graphql::{http::RawRequest, schema::Schema};
use open_dds::session_variables::{SessionVariable, SESSION_VARIABLE_ROLE};
use serde_json as json;
use std::{
    collections::HashMap,
    fs::{self, File},
    io::Write,
    path::Path,
    path::PathBuf,
};

use engine::execute::{execute_query, HttpContext};
use engine::schema::GDS;

extern crate json_value_merge;
use json_value_merge::Merge;
use serde_json::Value;

pub struct GoldenTestContext {
    http_context: HttpContext,
    mint: Mint,
}

pub fn setup(test_dir: &Path) -> GoldenTestContext {
    let http_context = HttpContext {
        client: reqwest::Client::new(),
        ndc_response_size_limit: None,
    };
    let mint = Mint::new(test_dir);
    GoldenTestContext { http_context, mint }
}

fn resolve_session(
    session_variables: HashMap<SessionVariable, SessionVariableValue>,
) -> Result<Session, SessionError> {
    //return an arbitrary identity with role emulation enabled
    let authorization = Identity::admin(Role::new("admin"));
    let role = session_variables
        .get(&SESSION_VARIABLE_ROLE)
        .map(|v| Role::new(&v.0));
    let role_authorization = authorization.get_role_authorization(role.as_ref())?;
    let session = role_authorization.build_session(session_variables);
    Ok(session)
}

// This function is deprecated in favour of test_execution_expectation
// TODO: Remove this function after all tests are moved to use test_execution_expectation
pub fn test_execution_expectation_legacy(
    test_path_string: &str,
    common_metadata_paths: &[&str],
) -> anyhow::Result<()> {
    tokio_test::block_on(async {
        // Setup test context
        let root_test_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("tests");
        let mut test_ctx = setup(&root_test_dir);
        let test_path = root_test_dir.join(test_path_string);

        let request_path = test_path.join("request.gql");
        let response_path = test_path_string.to_string() + "/expected.json";
        let metadata_path = test_path.join("metadata.json");

        let metadata_json_value = merge_with_common_metadata(
            &metadata_path,
            common_metadata_paths
                .iter()
                .map(|path| root_test_dir.join(path)),
        )?;

        let metadata = open_dds::traits::OpenDd::deserialize(metadata_json_value)?;

        // TODO: remove this assert once we have stopped manually implementing Serialize for OpenDD types.
        assert_eq!(
            open_dds::Metadata::from_json_str(&serde_json::to_string(&metadata)?)?,
            metadata
        );

        let gds = GDS::new(metadata)?;
        let schema = GDS::build_schema(&gds)?;

        // Ensure schema is serialized successfully.
        serde_json::to_string(&schema)?;

        let query = fs::read_to_string(request_path)?;

        let session = {
            let session_vars_path = &test_path.join("session_variables.json");
            let session_variables: HashMap<SessionVariable, SessionVariableValue> =
                json::from_str(fs::read_to_string(session_vars_path)?.as_ref())?;
            resolve_session(session_variables)
        }?;

        let raw_request = RawRequest {
            operation_name: None,
            query,
            variables: None,
        };

        // Execute the test

        let response =
            execute_query(&test_ctx.http_context, &schema, &session, raw_request, None).await;

        let mut expected = test_ctx.mint.new_goldenfile_with_differ(
            response_path,
            Box::new(|file1, file2| {
                let json1: serde_json::Value =
                    serde_json::from_reader(File::open(file1).unwrap()).unwrap();
                let json2: serde_json::Value =
                    serde_json::from_reader(File::open(file2).unwrap()).unwrap();
                if json1 != json2 {
                    text_diff(file1, file2)
                }
            }),
        )?;
        write!(expected, "{}", serde_json::to_string_pretty(&response.0)?)?;
        Ok(())
    })
}

#[allow(dead_code)]
pub(crate) fn test_introspection_expectation(
    test_path_string: &str,
    common_metadata_paths: &[&str],
) -> anyhow::Result<()> {
    tokio_test::block_on(async {
        // Setup test context
        let root_test_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("tests");
        let mut test_ctx = setup(&root_test_dir);
        let test_path = root_test_dir.join(test_path_string);

        let request_path = test_path.join("introspection_request.gql");
        let response_path = test_path_string.to_string() + "/introspection_expected.json";
        let metadata_path = test_path.join("metadata.json");

        let metadata_json_value = merge_with_common_metadata(
            &metadata_path,
            common_metadata_paths
                .iter()
                .map(|path| root_test_dir.join(path)),
        )?;

        let metadata = open_dds::traits::OpenDd::deserialize(metadata_json_value)?;

        // TODO: remove this assert once we have stopped manually implementing Serialize for OpenDD types.
        assert_eq!(
            open_dds::Metadata::from_json_str(&serde_json::to_string(&metadata)?)?,
            metadata
        );

        let gds = GDS::new(metadata)?;
        let schema = GDS::build_schema(&gds)?;

        // Verify successful serialization and deserialization of the schema.
        // Hasura V3 relies on the serialized schema for handling requests.
        // Therefore, it is crucial to ensure the functionality of both
        // deserialization and serialization.
        // Testing this within this function allows us to detect errors for any
        // future metadata tests that may be added.
        let serialized_metadata =
            serde_json::to_string(&schema).expect("Failed to serialize schema");
        let deserialized_metadata: Schema<GDS> =
            serde_json::from_str(&serialized_metadata).expect("Failed to deserialize metadata");
        assert_eq!(
            schema, deserialized_metadata,
            "initial built metadata does not match deserialized metadata"
        );

        let query = fs::read_to_string(request_path)?;

        let session_vars_path = &test_path.join("session_variables.json");
        let sessions: Vec<HashMap<SessionVariable, SessionVariableValue>> =
            json::from_str(fs::read_to_string(session_vars_path)?.as_ref())?;
        let sessions: Vec<Session> = sessions
            .into_iter()
            .map(resolve_session)
            .collect::<Result<_, _>>()?;

        assert!(
            sessions.len() > 1,
            "Found less than 2 roles in test scenario"
        );

        let raw_request = RawRequest {
            operation_name: None,
            query,
            variables: None,
        };

        // Execute the test
        let mut responses = Vec::new();
        for session in sessions.iter() {
            let response = execute_query(
                &test_ctx.http_context,
                &schema,
                session,
                raw_request.clone(),
                None,
            )
            .await;
            responses.push(response.0);
        }

        let mut expected = test_ctx.mint.new_goldenfile_with_differ(
            response_path,
            Box::new(|file1, file2| {
                let json1: serde_json::Value =
                    serde_json::from_reader(File::open(file1).unwrap()).unwrap();
                let json2: serde_json::Value =
                    serde_json::from_reader(File::open(file2).unwrap()).unwrap();
                if json1 != json2 {
                    text_diff(file1, file2)
                }
            }),
        )?;
        write!(expected, "{}", serde_json::to_string_pretty(&responses)?)?;
        Ok(())
    })
}

pub fn test_execution_expectation(
    test_path_string: &str,
    common_metadata_paths: &[&str],
) -> anyhow::Result<()> {
    tokio_test::block_on(async {
        // Setup test context
        let root_test_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("tests");
        let mut test_ctx = setup(&root_test_dir);
        let test_path = root_test_dir.join(test_path_string);

        let request_path = test_path.join("request.gql");
        let response_path = test_path_string.to_string() + "/expected.json";
        let metadata_path = test_path.join("metadata.json");

        let metadata_json_value = merge_with_common_metadata(
            &metadata_path,
            common_metadata_paths
                .iter()
                .map(|path| root_test_dir.join(path)),
        )?;

        let metadata = open_dds::traits::OpenDd::deserialize(metadata_json_value)?;

        // TODO: remove this assert once we have stopped manually implementing Serialize for OpenDD types.
        assert_eq!(
            open_dds::Metadata::from_json_str(&serde_json::to_string(&metadata)?)?,
            metadata
        );

        let gds = GDS::new(metadata)?;
        let schema = GDS::build_schema(&gds)?;

        // Verify successful serialization and deserialization of the schema.
        // Hasura V3 relies on the serialized schema for handling requests.
        // Therefore, it is crucial to ensure the functionality of both
        // deserialization and serialization.
        // Testing this within this function allows us to detect errors for any
        // future metadata tests that may be added.
        let serialized_metadata =
            serde_json::to_string(&schema).expect("Failed to serialize schema");
        let deserialized_metadata: Schema<GDS> =
            serde_json::from_str(&serialized_metadata).expect("Failed to deserialize metadata");
        assert_eq!(
            schema, deserialized_metadata,
            "initial built metadata does not match deserialized metadata"
        );

        let query = fs::read_to_string(request_path)?;

        let session_vars_path = &test_path.join("session_variables.json");
        let sessions: Vec<HashMap<SessionVariable, SessionVariableValue>> =
            json::from_str(fs::read_to_string(session_vars_path)?.as_ref())?;
        let sessions: Vec<Session> = sessions
            .into_iter()
            .map(resolve_session)
            .collect::<Result<_, _>>()?;

        assert!(
            sessions.len() > 1,
            "Found less than 2 roles in test scenario"
        );

        let raw_request = RawRequest {
            operation_name: None,
            query,
            variables: None,
        };

        // Execute the test
        let mut responses = Vec::new();
        for session in sessions.iter() {
            let response = execute_query(
                &test_ctx.http_context,
                &schema,
                session,
                raw_request.clone(),
                None,
            )
            .await;
            responses.push(response.0);
        }

        let mut expected = test_ctx.mint.new_goldenfile_with_differ(
            response_path,
            Box::new(|file1, file2| {
                let json1: serde_json::Value =
                    serde_json::from_reader(File::open(file1).unwrap()).unwrap();
                let json2: serde_json::Value =
                    serde_json::from_reader(File::open(file2).unwrap()).unwrap();
                if json1 != json2 {
                    text_diff(file1, file2)
                }
            }),
        )?;
        write!(expected, "{}", serde_json::to_string_pretty(&responses)?)?;
        Ok(())
    })
}

fn read_json(path: &Path) -> anyhow::Result<Value> {
    let json_string = fs::read_to_string(path)?;
    let value = serde_json::from_str(&json_string)?;
    Ok(value)
}

pub fn merge_with_common_metadata<T: Iterator<Item = PathBuf>>(
    metadata_path: &Path,
    common_metadata_paths: T,
) -> anyhow::Result<Value> {
    let mut metadata = read_json(metadata_path)?;
    for path in common_metadata_paths {
        let common_metadata = read_json(&path)?;
        metadata.merge(&common_metadata);
    }
    Ok(metadata)
}

#[allow(dead_code)]
pub fn test_execute_explain(
    test_path_string: &str,
    test_metadata_path: &str,
    common_metadata_paths: &[&str],
) -> anyhow::Result<()> {
    tokio_test::block_on(async {
        let root_test_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("tests");
        let mut test_ctx = setup(&root_test_dir);
        let test_path = root_test_dir.join(test_path_string);
        let gql_request_file_path = test_path.join("request.gql");
        let expected_response_file = test_path_string.to_string() + "/expected.json";

        let test_metadata_path = root_test_dir.join(test_metadata_path);
        let metadata = merge_with_common_metadata(
            &test_metadata_path,
            common_metadata_paths
                .iter()
                .map(|path| root_test_dir.join(path)),
        )?;
        let gds = GDS::new(open_dds::traits::OpenDd::deserialize(metadata)?)?;
        let schema = GDS::build_schema(&gds)?;
        let session = {
            let session_variables_raw = r#"{
                "x-hasura-role": "admin"
            }"#;
            let session_variables: HashMap<SessionVariable, SessionVariableValue> =
                serde_json::from_str(session_variables_raw)?;
            resolve_session(session_variables)
        }?;
        let query = std::fs::read_to_string(root_test_dir.join(gql_request_file_path))?;
        let raw_request = lang_graphql::http::RawRequest {
            operation_name: None,
            query,
            variables: None,
        };
        let raw_response = engine::execute::explain::execute_explain(
            &test_ctx.http_context,
            &schema,
            &session,
            raw_request,
        )
        .await;

        let response = engine::execute::explain::types::redact_ndc_explain(raw_response);

        let mut expected = test_ctx.mint.new_goldenfile_with_differ(
            expected_response_file,
            Box::new(|file1, file2| {
                let json1: serde_json::Value =
                    serde_json::from_reader(File::open(file1).unwrap()).unwrap();
                let json2: serde_json::Value =
                    serde_json::from_reader(File::open(file2).unwrap()).unwrap();
                if json1 != json2 {
                    text_diff(file1, file2)
                }
            }),
        )?;
        write!(expected, "{}", serde_json::to_string_pretty(&response)?)?;
        Ok(())
    })
}
