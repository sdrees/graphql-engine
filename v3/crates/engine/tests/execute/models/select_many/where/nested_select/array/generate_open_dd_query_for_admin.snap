---
source: crates/engine/tests/common.rs
expression: ir
---
V1(
    QueryRequestV1 {
        queries: {
            Alias(
                Identifier(
                    "where_does_john_hughes_work",
                ),
            ): Model(
                ModelSelection {
                    target: ModelTarget {
                        subgraph: SubgraphName(
                            "default",
                        ),
                        model_name: ModelName(
                            Identifier(
                                "institutions",
                            ),
                        ),
                        arguments: {
                            ArgumentName(
                                Identifier(
                                    "where",
                                ),
                            ): Literal(
                                Object {
                                    "staff": Object {
                                        "last_name": Object {
                                            "_eq": String("Hughes"),
                                        },
                                    },
                                },
                            ),
                        },
                        filter: None,
                        order_by: [],
                        limit: None,
                        offset: None,
                    },
                    selection: {
                        Alias(
                            Identifier(
                                "id",
                            ),
                        ): Field(
                            ObjectFieldSelection {
                                target: ObjectFieldTarget {
                                    field_name: FieldName(
                                        Identifier(
                                            "id",
                                        ),
                                    ),
                                    arguments: {},
                                },
                                selection: None,
                            },
                        ),
                        Alias(
                            Identifier(
                                "location",
                            ),
                        ): Field(
                            ObjectFieldSelection {
                                target: ObjectFieldTarget {
                                    field_name: FieldName(
                                        Identifier(
                                            "location",
                                        ),
                                    ),
                                    arguments: {},
                                },
                                selection: Some(
                                    {
                                        Alias(
                                            Identifier(
                                                "city",
                                            ),
                                        ): Field(
                                            ObjectFieldSelection {
                                                target: ObjectFieldTarget {
                                                    field_name: FieldName(
                                                        Identifier(
                                                            "city",
                                                        ),
                                                    ),
                                                    arguments: {},
                                                },
                                                selection: None,
                                            },
                                        ),
                                        Alias(
                                            Identifier(
                                                "campuses",
                                            ),
                                        ): Field(
                                            ObjectFieldSelection {
                                                target: ObjectFieldTarget {
                                                    field_name: FieldName(
                                                        Identifier(
                                                            "campuses",
                                                        ),
                                                    ),
                                                    arguments: {},
                                                },
                                                selection: None,
                                            },
                                        ),
                                    },
                                ),
                            },
                        ),
                    },
                },
            ),
            Alias(
                Identifier(
                    "where_has_staff_with_a_dog",
                ),
            ): Model(
                ModelSelection {
                    target: ModelTarget {
                        subgraph: SubgraphName(
                            "default",
                        ),
                        model_name: ModelName(
                            Identifier(
                                "institutions",
                            ),
                        ),
                        arguments: {
                            ArgumentName(
                                Identifier(
                                    "where",
                                ),
                            ): Literal(
                                Object {
                                    "staff": Object {
                                        "pets": Object {
                                            "name": Object {
                                                "_eq": String("Dog"),
                                            },
                                        },
                                    },
                                },
                            ),
                        },
                        filter: None,
                        order_by: [],
                        limit: None,
                        offset: None,
                    },
                    selection: {
                        Alias(
                            Identifier(
                                "id",
                            ),
                        ): Field(
                            ObjectFieldSelection {
                                target: ObjectFieldTarget {
                                    field_name: FieldName(
                                        Identifier(
                                            "id",
                                        ),
                                    ),
                                    arguments: {},
                                },
                                selection: None,
                            },
                        ),
                        Alias(
                            Identifier(
                                "location",
                            ),
                        ): Field(
                            ObjectFieldSelection {
                                target: ObjectFieldTarget {
                                    field_name: FieldName(
                                        Identifier(
                                            "location",
                                        ),
                                    ),
                                    arguments: {},
                                },
                                selection: Some(
                                    {
                                        Alias(
                                            Identifier(
                                                "city",
                                            ),
                                        ): Field(
                                            ObjectFieldSelection {
                                                target: ObjectFieldTarget {
                                                    field_name: FieldName(
                                                        Identifier(
                                                            "city",
                                                        ),
                                                    ),
                                                    arguments: {},
                                                },
                                                selection: None,
                                            },
                                        ),
                                        Alias(
                                            Identifier(
                                                "campuses",
                                            ),
                                        ): Field(
                                            ObjectFieldSelection {
                                                target: ObjectFieldTarget {
                                                    field_name: FieldName(
                                                        Identifier(
                                                            "campuses",
                                                        ),
                                                    ),
                                                    arguments: {},
                                                },
                                                selection: None,
                                            },
                                        ),
                                    },
                                ),
                            },
                        ),
                    },
                },
            ),
            Alias(
                Identifier(
                    "institution_in_a_country_with_a_city_named_stockholm",
                ),
            ): Model(
                ModelSelection {
                    target: ModelTarget {
                        subgraph: SubgraphName(
                            "default",
                        ),
                        model_name: ModelName(
                            Identifier(
                                "institutions",
                            ),
                        ),
                        arguments: {
                            ArgumentName(
                                Identifier(
                                    "where",
                                ),
                            ): Literal(
                                Object {
                                    "location": Object {
                                        "country": Object {
                                            "cities": Object {
                                                "name": Object {
                                                    "_eq": String("Stockholm"),
                                                },
                                            },
                                        },
                                    },
                                },
                            ),
                        },
                        filter: None,
                        order_by: [],
                        limit: None,
                        offset: None,
                    },
                    selection: {
                        Alias(
                            Identifier(
                                "id",
                            ),
                        ): Field(
                            ObjectFieldSelection {
                                target: ObjectFieldTarget {
                                    field_name: FieldName(
                                        Identifier(
                                            "id",
                                        ),
                                    ),
                                    arguments: {},
                                },
                                selection: None,
                            },
                        ),
                        Alias(
                            Identifier(
                                "location",
                            ),
                        ): Field(
                            ObjectFieldSelection {
                                target: ObjectFieldTarget {
                                    field_name: FieldName(
                                        Identifier(
                                            "location",
                                        ),
                                    ),
                                    arguments: {},
                                },
                                selection: Some(
                                    {
                                        Alias(
                                            Identifier(
                                                "country",
                                            ),
                                        ): Field(
                                            ObjectFieldSelection {
                                                target: ObjectFieldTarget {
                                                    field_name: FieldName(
                                                        Identifier(
                                                            "country",
                                                        ),
                                                    ),
                                                    arguments: {},
                                                },
                                                selection: Some(
                                                    {
                                                        Alias(
                                                            Identifier(
                                                                "cities",
                                                            ),
                                                        ): Field(
                                                            ObjectFieldSelection {
                                                                target: ObjectFieldTarget {
                                                                    field_name: FieldName(
                                                                        Identifier(
                                                                            "cities",
                                                                        ),
                                                                    ),
                                                                    arguments: {},
                                                                },
                                                                selection: Some(
                                                                    {
                                                                        Alias(
                                                                            Identifier(
                                                                                "name",
                                                                            ),
                                                                        ): Field(
                                                                            ObjectFieldSelection {
                                                                                target: ObjectFieldTarget {
                                                                                    field_name: FieldName(
                                                                                        Identifier(
                                                                                            "name",
                                                                                        ),
                                                                                    ),
                                                                                    arguments: {},
                                                                                },
                                                                                selection: None,
                                                                            },
                                                                        ),
                                                                    },
                                                                ),
                                                            },
                                                        ),
                                                    },
                                                ),
                                            },
                                        ),
                                    },
                                ),
                            },
                        ),
                    },
                },
            ),
        },
    },
)
